pkg/controller/admissionchecks/multikueue/fswatch_test.go

```diff
@@ -0,0 +1,460 @@
+			},
+			wantEventsForClusters: set.New("c1"),
+		},
+		"single cluster file rename from": {
+			prepareFnc: func(basePath string) error {
+				return os.WriteFile(path.Join(basePath, "c1.kubeconfig"), []byte("123"), 0666)
+			},
+			clusters: map[string]string{
+				"c1": "c1.kubeconfig",
+			},
+			opFnc: func(basePath string) error {
+				return os.Rename(path.Join(basePath, "c1.kubeconfig"), path.Join(basePath, "c1.kubeconfig.new"))
+			},
+			wantEventsForClusters: set.New("c1"),
+		},
+		"single crete link": {
+			prepareFnc: func(basePath string) error {
+				return os.WriteFile(path.Join(basePath, "c1.kubeconfig.src"), []byte("123"), 0666)
+			},
+			clusters: map[string]string{
+				"c1": "c1.kubeconfig",
+			},
+			opFnc: func(basePath string) error {
+				return os.Symlink(path.Join(basePath, "c1.kubeconfig.src"), path.Join(basePath, "c1.kubeconfig"))
+			},
+			wantEventsForClusters: set.New("c1"),
+		},
+		"single remove link": {
+			prepareFnc: func(basePath string) error {
+				if err := os.WriteFile(path.Join(basePath, "c1.kubeconfig.src"), []byte("123"), 0666); err != nil {
+					return err
+				}
+				return os.Symlink(path.Join(basePath, "c1.kubeconfig.src"), path.Join(basePath, "c1.kubeconfig"))
+			},
+			clusters: map[string]string{
+				"c1": "c1.kubeconfig",
+			},
+			opFnc: func(basePath string) error {
+				return os.Remove(path.Join(basePath, "c1.kubeconfig"))
+			},
+			wantEventsForClusters: set.New("c1"),
+		},
+	}
+
+	for name, tc := range cases {
+		t.Run(name, func(t *testing.T) {
+			basePath := t.TempDir()
+			if tc.prepareFnc != nil {
+				if err := tc.prepareFnc(basePath); err != nil {
+					t.Fatalf("unexpected prepare error: %s", err)
+				}
+			}
+			ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+			watcher := newKubeConfigFSWatcher()
+
+			// start the recorder
+			gotEventsForClusters := set.New[string]()
+			wg := sync.WaitGroup{}
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+				for {
+					select {
+					case ev := <-watcher.reconcile:
+						gotEventsForClusters.Insert(ev.Object.GetName())
+						if len(gotEventsForClusters) == len(tc.wantEventsForClusters) {
+							cancel()
+						}
+					case <-ctx.Done():
+						return
+					}
+				}
+			}()
+			_ = watcher.Start(ctx)
+
+			gotAddErrors := map[string]error{}
+			for c, p := range tc.clusters {
+				err := watcher.AddOrUpdate(c, path.Join(basePath, p))
+				if err != nil {
+					gotAddErrors[c] = err
+				}
+			}
+
+			if tc.opFnc != nil {
+				if err := tc.opFnc(basePath); err != nil {
+					t.Fatalf("unexpected op error: %s", err)
+				}
+			}
+			// wait
+			wg.Wait()
+
+			if len(gotAddErrors) > 0 {
+				t.Errorf("unexpected add errors:%s", gotAddErrors)
+			}
+
+			if diff := cmp.Diff(tc.wantEventsForClusters, gotEventsForClusters); diff != "" {
+				t.Errorf("unexpected events for clusters(-want/+got):\n%s", diff)
+			}
+
+		})
+	}
+}
+
+func TestFSWatchAddRm(t *testing.T) {
```

Comment:
This is a lot of unit testing. Do they cover 100% the new code?

Reply:
sgtm, so we have unit tests focused on the watcher and state management, and an integration test to make sure they talk to each other.